subprojects/core/src/test/groovy/org/gradle/internal/vfs/AbstractFileWatcherUpdaterTest.groovy

```diff
@@ -0,0 +1,133 @@
+
+    def "starts and stops watching must watch directories"() {
+        def mustWatchDirectories = ["first", "second", "third"].collect { temporaryFolder.testDirectory.createDir(it)}
+
+        when:
+        updater.updateMustWatchDirectories(mustWatchDirectories)
+        then:
+        1 * watcher.startWatching({ equalAsSets(it, mustWatchDirectories) })
+        0 * _
+
+        when:
+        updater.updateMustWatchDirectories([])
+        then:
+        1 * watcher.stopWatching({ equalAsSets(it, mustWatchDirectories) })
+        0 * _
+    }
+
+    def "does not watch non-existing must watch directories"() {
+        def existingMustWatchDirectories = ["first", "second"].collect { file(it).createDir() }
+        def nonExistingMustWatchDirectory = file("non-existing")
+
+        when:
+        updater.updateMustWatchDirectories(existingMustWatchDirectories + nonExistingMustWatchDirectory)
+        then:
+        1 * watcher.startWatching({ equalAsSets(it, existingMustWatchDirectories) })
+        0 * _
+    }
+
+    def "can change the must watch directories"() {
+        def firstMustWatchDirectories = ["first", "second"].collect { file(it).createDir() }
+        def secondMustWatchDirectories = ["second", "third"].collect { file(it).createDir() }
+
+        when:
+        updater.updateMustWatchDirectories(firstMustWatchDirectories)
+        then:
+        1 * watcher.startWatching({ equalAsSets(it, firstMustWatchDirectories) })
+        0 * _
+
+        when:
+        updater.updateMustWatchDirectories(secondMustWatchDirectories)
+        then:
+        1 * watcher.stopWatching({ equalAsSets(it, [file("first")]) })
+        then:
+        1 * watcher.startWatching({ equalAsSets(it, [file("third")]) })
+        0 * _
+    }
+
+    TestFile file(Object... path) {
+        temporaryFolder.testDirectory.file(path)
+    }
+
+    CompleteDirectorySnapshot snapshotDirectory(File directory) {
+        directorySnapshotter.snapshot(directory.absolutePath, null, new AtomicBoolean(false)) as CompleteDirectorySnapshot
+    }
+
+    void addSnapshot(CompleteFileSystemLocationSnapshot snapshot) {
+        root = decorator.decorate({ currentRoot, listener -> currentRoot.store(snapshot.absolutePath, snapshot, listener) }).updateRoot(root)
+    }
+
+    void invalidate(String absolutePath) {
+        root = decorator.decorate({ currentRoot, listener -> currentRoot.invalidate(absolutePath, listener) }).updateRoot(root)
+    }
+
+    void invalidate(CompleteFileSystemLocationSnapshot snapshot) {
+        invalidate(snapshot.absolutePath)
+    }
+
+    static RegularFileSnapshot snapshotRegularFile(File regularFile) {
+        new RegularFileSnapshot(regularFile.absolutePath, regularFile.name, TestFiles.fileHasher().hash(regularFile), FileMetadata.from(Files.readAttributes(regularFile.toPath(), BasicFileAttributes)))
+    }
+
+    static boolean equalAsSets(Object actual, Object expected) {
+        return (actual as Set) == (expected as Set)
+    }
```

Comment:
This is problematic. I think we should be able to declare the type of the parameters more strictly.

I would also prefer if instead of converting to `Set` we'd compare a sorted list of elements. That way we'd be able to spot if an element appears multiple times (which it shouldn't).

Reply:
I used a collection first, though then IDEA was not able to infer the argument types from where this method has been used. That is why I ended up with `Object` in the end...

Reply:
Now the second argument is a `Collection` and the method compares the sorted collections.